### PR TITLE
udev: modernize code and propagate critical errors

### LINKED ISSUE #12291: udev: RUN assignments and PROGRAM results
This bug report is based on the latest sourcecode of udev, I had a really hard time to debug my [udev rule](https://github.com/atar-axis/xpadneo/issues/91) and therefore had a closer look into the sourcecode.

**The first set of issues is about the `RUN` key**

Problematic code: [here](https://github.com/systemd/systemd/blob/15de23a0b24f6073d527c6a8c22556ad5101a160/src/udev/udev-rules.c#L2442)

1. <strike>Final assignments on the `RUN` key cannot work. This is because `event->run_final` exists, but is never set or checked. I therefore suggest to add the following:
    ```
    if (event->run_final)
        break;
    if (cur->key.op == OP_ASSIGN_FINAL)
        event->run_final = true;
    ```
    If this behaviour is intended, then please document it :blush:
   </strike> (fixed by [PR](https://github.com/systemd/systemd/pull/12309))

2. <strike>Using `=` or `:=` on the `RUN{builtin}` key does clear the whole `event->run_list`, even `RUN{program}` entries (and vice versa). This is not documented either. I suggest to separate both `RUN` commands or documenting this behaviour.</strike> (fixed by [PR](https://github.com/systemd/systemd/pull/12310))

3. <strike>It should be documented that `RUN` commands are not executed sequentially</strike> (fixed by https://github.com/systemd/systemd/pull/11891/commits/541d8bd4b18a4b97a381ae6caea4aba2baaa063c)

**The next issue is about the `PROGRAM` and `RESULT` keys**

Imagine the following situation:
* You have two different rules, both fire on the same uevent.
* Both rules have an `PROGRAM` command and both are using the `$result` variable in their `RUN` commands.

Now the problem is, that the `PROGRAM` commands are executed before the `RUN` commands, meaning that the second rule is overriding the `$result` of the first one.
Therefore the `RUN` command of the first rule does now have the `$result` of `PROGRAM` of the second rule.

I suggest to evaluate `PROGRAM` first and replacing the `$result` variable in `RUN` commands before it gets added to the `event->run_list`. At the moment the substitution does happen shortly before execution, see [here](https://github.com/systemd/systemd/blob/3661dc349ec0e0e8f19fb3a542ccbc813753ef8b/src/udev/udev-event.c#L907). Otherwise, please document this behaviour! :relaxed:

### PR DESCRIPTION
**Pull request by yuwata**
Replaces #12416. Closes #12291.

@keszybz I cannot split the 3rd commit. Sorry. Could you take a look?

### CODE REVIEW

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -2311,5 +2343,26 @@ def test_dhcp_client_route_remove_on_renew(self):
         self.assertRegex(output, f'192.168.5.1 proto dhcp scope link src {address2} metric 1024')
 
 if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--build-dir', help='Path to build dir', dest='build_dir')
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: No need to specify `dest` here, `build_dir` will be used automatically.

File: test/test-network/systemd-networkd-tests.py [hunk 2]
```diff
@@ -2311,5 +2343,26 @@ def test_dhcp_client_route_remove_on_renew(self):
...
+    if ns.build_dir:
+        networkd_bin = os.path.join(ns.build_dir, 'systemd-networkd')
+        wait_online_bin = os.path.join(ns.build_dir, 'systemd-networkd-wait-online')
+        networkctl_bin = os.path.join(ns.build_dir, 'networkctl')
+    else:
+        networkd_bin = ns.networkd_bin
+        wait_online_bin = ns.wait_online_bin
+        networkctl_bin = ns.networkctl_bin
+    use_valgrind = ns.use_valgrind
+    enable_debug = ns.enable_debug
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py]: <strike>
This seems overly complicated. Also, an explicit argument like `--networkctl` will be ignored if `--build-dir` is specified.
What about the following:
```
parser = argparse.ArgumentParser()
parser.add_argument('--build-dir', help='Path to build dir')
parser.add_argument('--networkd', help='Path to systemd-networkd')
...
</strike>

I"ll submit a separate patch later. I think it'll be quicker then me trying to describe all the changes I'd like to see.

File: src/basic/fs-util.h [hunk 3]
```diff
@@ -14,6 +14,13 @@
 #include "errno-util.h"
 #include "time-util.h"
 
+#define MODE_INVALID ((mode_t) -1)
+
+/* The following macros add 1 when converting things, since 0 is a valid mode, while the pointer
+ * NULL is special */
+#define PTR_TO_MODE(p) ((mode_t) ((uintptr_t) (p)-1))
+#define MODE_TO_PTR(u) ((void *) ((uintptr_t) (u)+1))
```
Comment (keszybz) [inline src/basic/fs-util.h:22]: Shouldn't MODE_INVALID be handled specially, i.e. not have +1 added?

File: src/udev/udev-event.c [hunk 4]
```diff
@@ -753,19 +758,23 @@ static int update_devnode(UdevEvent *event) {
         if (event->dev_db_clone)
                 (void) udev_node_update_old_links(dev, event->dev_db_clone);
 
-        if (!event->owner_set) {
+        if (!uid_is_valid(event->uid)) {
                 r = device_get_devnode_uid(dev, &event->uid);
                 if (r < 0 && r != -ENOENT)
                         return log_device_error_errno(dev, r, "Failed to get devnode UID: %m");
+                if (r == -ENOENT)
+                        event->uid = 0;
```
Comment (keszybz) [inline src/udev/udev-event.c:823]: Maybe
```C
if (r == -ENOENT)
       ...
else if (r < 0)
      ...
```
(Not that it matters.)

File: src/udev/udev-rules.c [hunk 5]
```diff
@@ -27,1441 +15,911 @@
...
-
-static char *rules_str(UdevRules *rules, unsigned off) {
-        return rules->strbuf->buf + off;
-}
-
-static unsigned rules_add_string(UdevRules *rules, const char *s) {
-        return strbuf_add_string(rules->strbuf, s, strlen(s));
+static void udev_rule_token_free(UdevRuleToken *token) {
+        free(token);
 }
```
Comment (keszybz) [inline src/udev/udev-rules.c:35]: Let's kill that too, but maybe later.

File: src/udev/udev-event.c [hunk 6]
```diff
@@ -134,6 +134,85 @@ static const struct subst_map_entry map[] = {
            { .name = "sys",      .fmt = 'S', .type = FORMAT_SUBST_SYS },
 };
 
+static const char *format_type_to_string(FormatSubstitutionType t) {
+        size_t i;
+
+        for (i = 0; i < ELEMENTSOF(map); i++)
```
Comment (keszybz) [inline src/udev/udev-event.c]: I'd just put `size_t ` here to save two lines...

File: src/udev/udev-event.c [hunk 7]
```diff
@@ -213,6 +213,39 @@ static int get_subst_type(const char **str, FormatSubstitutionType *ret_type, ch
...
+                has_plus = false;
+        else if (x[0] == '+' && x[1] == '\0')
+                has_plus = true;
+        else
+                return -EINVAL;
+
+        *ret_u = (unsigned) l;
+        *ret_has_plus = has_plus;
+        return 0;
+}
```
Comment (keszybz) [inline src/udev/udev-event.c:228]: Hmm, what about the following instead:
```C
static int safe_atou_optional_plus(const char *s, unsigned *ret_u) {
    /* Returns 1 if plus, 0 if no plus, negative on error */
    int r;
    const char *p = endswith(s, "+");
    if (p)
            s = strndupa(s, p - s);
    r =  safe_atou(s, ret_u);
    if (r < 0)
           return r;
    return !!p;
}
```
(or keep `*ret_has_plus`, whatever is more convenient).

File: src/udev/udev-rules.c [hunk 8]
```diff
@@ -1146,15 +1148,22 @@ static bool token_match_attr(UdevRuleToken *token, sd_device *dev, UdevEvent *ev
...
+                        return (int) ret;
                 name = nbuf;
                 _fallthrough_;
         case SUBST_TYPE_PLAIN:
-                if (sd_device_get_sysattr_value(dev, name, &value) < 0)
+                r = sd_device_get_sysattr_value(dev, name, &value);
+                if (r == -ENOENT)
+                        /* always false even if value is empty and operator is '==', e.g., ATTR{hoge}=="" */
                         return false;
+                if (r < 0 && r == -ENOENT)
```
Comment (keszybz) [inline src/udev/udev-rules.c]: The right part of the conditional cannot be satifsfied.

File: src/udev/udev-event.c [hunk 9]
```diff
@@ -994,8 +994,13 @@ void udev_event_execute_run(UdevEvent *event, usec_t timeout_usec) {
         ORDERED_HASHMAP_FOREACH_KEY(val, cmd, event->run_list, i) {
                 UdevBuiltinCommand builtin_cmd = PTR_TO_UDEV_BUILTIN_CMD(val);
                 char command[UTIL_PATH_SIZE];
+                ssize_t ret;
 
-                (void) udev_event_apply_format(event, cmd, command, sizeof(command), false);
+                ret = udev_event_apply_format(event, cmd, command, sizeof(command), false);
+                if (ret < 0) {
+                        log_device_debug_errno(event->dev, (int) ret, "Failed to apply format to '%s', skipping to execute command: %m", cmd);
```
Comment (keszybz) [inline src/udev/udev-event.c]: Grammar seems a bit off here. Maybe `"Failed to format \"%s\", skipping command: %m"` ?

**Comment by keszybz** [comment]
Is the first patch here by mistake? I see it also in #12699.

**Comment by yuwata** [comment]
> Is the first patch here by mistake? I see it also in #12699.

Seems so. Sorry. Will drop.

File: src/basic/fs-util.h (hunk 3 above)
Comment (yuwata) [reply src/basic/fs-util.h:22]: No, here +1 or -1 intends NULL <-> MODE_INVALID, that is, non-NULL means valid mode.
We already do the same thing in `UID_TO_PTR()`/`PTR_TO_UID()`.

File: src/udev/udev-rules.c (hunk 5 above)
Comment (yuwata) [reply src/udev/udev-rules.c:35]: Personally, I'd like to keep this...

**Comment by yuwata** [comment]
@keszybz Thank you for your review. I've force-pushed an updated version. Most of your comments are addressed. Exceptions are `MODE_TO_PTR` and `udev_rule_token_free()` things. PTAL.

**Comment by yuwata** [comment]
@mrc0mmand Is it possible to obtain udevd's logs during running systemd-networkd-tests.py? It seems some lines are ignored...

**Comment by mrc0mmand** [comment]
@yuwata will the Arch VM's journal help? https://ci.centos.org/job/systemd-pr-build/6666/artifact//systemd-centos-ci/artifacts_3_e2Yh/testsuite-logs-upstream.ner/journald-testsuite_PASS.log (it's the journal from the Arch KVM machine, not from the QEMU VMs for `test/TEST-??-*`)

https://ci.centos.org/job/systemd-pr-build/6666/artifact//systemd-centos-ci/index.html

Also, nice build ID :-)

**Comment by yuwata** [comment]
Thanks!
```
May 30 17:40:25 n3.pufty.ci.centos.org systemd-udevd[2827]: /usr/lib/udev/rules.d/84-nm-drivers.rules:10 Invalid value for PROGRAM.
May 30 17:40:25 n3.pufty.ci.centos.org systemd-udevd[2827]: /usr/lib/udev/rules.d/98-kexec.rules:1 The line takes no effect, ignoring.
May 30 17:40:25 n3.pufty.ci.centos.org systemd-udevd[2827]: /usr/lib/udev/rules.d/98-kexec.rules:2 The line takes no effect, ignoring.
May 30 17:40:25 n3.pufty.ci.centos.org systemd-udevd[2827]: /usr/lib/udev/rules.d/98-kexec.rules:3 The line takes no effect, ignoring.
May 30 17:40:25 n3.pufty.ci.centos.org systemd-udevd[2827]: /usr/lib/udev/rules.d/98-kexec.rules:4 The line takes no effect, ignoring.
```

> Also, nice build ID :-)

Haha :-)

**Comment by yuwata** [comment]
Hmm, the warnings seem irrelevant... And my laptop works fine with this PR...

**Comment by yuwata** [comment]
Hmm. @mrc0mmand Could you enable debug logging of `udevd`? Adding `udev.log_priority=debug` to kernel command line is enough.

**Comment by mrc0mmand** [comment]
@yuwata https://ci.centos.org/job/systemd-ci-build/632/artifact//artifacts_all/artifacts_5tCa5U/testsuite-logs-upstream.KsK/journalctl-testsuite_PASS.log

**Comment by yuwata** [comment]
Hmm...
```
Failed to run builtin 'hwdb --subsystem=pci': Invalid argument
```

**Comment by yuwata** [comment]
Ah, `hwdb.bin does not exist, please run 'systemd-hwdb update'`

**Comment by yuwata** [comment]
I force-pushed a revised version. I hope all CIs will be green. PTAL.

**Comment by yuwata** [comment]
Yey! All green now!

File: src/udev/udev-rules.c (hunk 5 above)
Comment (keszybz) [reply src/udev/udev-rules.c:35]: OK. Doesn't matter much either way.

**Comment by keszybz** [comment]
LGTM.

**Comment by haraldh** [comment]
This broke dracut in legacy network mode.

**Comment by haraldh** [comment]
https://github.com/systemd/systemd/issues/14935

**Pull request merged**